src/ILCompiler.Compiler/src/Compiler/PrecomputedMetadataManager.cs

```diff
@@ -0,0 +1,365 @@
+                        MethodDesc method = (MethodDesc)tse;
+                        metadataModules.Add(((MetadataType)method.OwningType).Module);
+                        methodMappings.Add(new MetadataMapping<MethodDesc>(method, metadataTokenValue));
+                    }
+                    else if (tse is FieldDesc)
+                    {
+                        FieldDesc field = (FieldDesc)tse;
+                        metadataModules.Add(((MetadataType)field.OwningType).Module);
+                        fieldMappings.Add(new MetadataMapping<FieldDesc>(field, metadataTokenValue));
+                    }
+                }
+                else
+                {
+                    int deadLdtokenSignifier = il.ReadLdcI4();
+                    Debug.Assert(deadLdtokenSignifier == 0);
+                    il.ReadLdcI4();
+                    il.ReadPop();
+                    il.ReadPop();
+                }
+            }
+        }
+
+        public static bool ModuleHasMetadataMappings(ModuleDesc metadataDescribingModule)
+        {
+            try
+            {
+                return metadataDescribingModule.GetTypeByCustomAttributeTypeName(MetadataMappingTypeName) != null;
+            }
+            catch { }
+
+            return false;
+        }
+
+        private MetadataLoadedInfo LoadMetadata()
+        {
+            HashSet<ModuleDesc> metadataModules = new HashSet<ModuleDesc>();
+            MetadataType typeWithMetadataMappings = (MetadataType)_metadataDescribingModule.GetTypeByCustomAttributeTypeName(MetadataMappingTypeName);
+
+            MethodDesc fullMetadataMethod = typeWithMetadataMappings.GetMethod("Metadata", null);
+            MethodDesc weakMetadataMethod = typeWithMetadataMappings.GetMethod("WeakMetadata", null);
+
+            ILProvider ilProvider = new ILProvider(null);
+
+            MetadataLoadedInfo result = new MetadataLoadedInfo();
+
+            if (fullMetadataMethod != null)
+            {
+                MethodIL fullMethodIL = ilProvider.GetMethodIL(fullMetadataMethod);
+                ReadMetadataMethod(fullMethodIL, out result.StrongTypeMappings, out result.MethodMappings, out result.FieldMappings, ref metadataModules);
+            }
+
+            if (weakMetadataMethod != null)
+            {
+                MethodIL weakMethodIL = ilProvider.GetMethodIL(weakMetadataMethod);
+                List<MetadataMapping<MethodDesc>> weakMethodMappings = new List<MetadataMapping<MethodDesc>>();
+                List<MetadataMapping<FieldDesc>> weakFieldMappings = new List<MetadataMapping<FieldDesc>>();
+                ReadMetadataMethod(weakMethodIL, out result.AllTypeMappings, out weakMethodMappings, out weakFieldMappings, ref metadataModules);
+                if ((weakMethodMappings.Count > 0) || (weakFieldMappings.Count > 0))
+                {
+                    // the format does not permit weak field/method mappings
+                    throw new BadImageFormatException();
```

Comment:
I generally believe that everything that is directly controlled by input should be an exception, but I can be convinced otherwise.